### ISSUE TITLE
Wrong address range when `mem_map` on MIPS architecture

### ISSUE THREAD

**Issue by chinggg** [unicorn-engine/unicorn#1708]
I happened to find a bug when using the Python binding of unicorn, not sure if it is a feature of MIPS architecture or it works differently in other language bindings.

```python
from unicorn import Uc
from unicorn.unicorn_const import *


mu = Uc(UC_ARCH_MIPS, UC_MODE_MIPS32 + UC_MODE_BIG_ENDIAN)  # MIPS64 and Little Endian have the same result
addr = 0x90000000
size = 0x2000
mu.mem_map(addr, size, UC_PROT_ALL)
for (s, e, p) in mu.mem_regions():
    print(f"{s:#x} - {e:#x} {p:#x}")
```

Expected output:
`0x90000000 - 0x90001fff 0x7`

Actual output:
`0x10000000 - 0x10001fff 0x7`

**Comment by PhilippTakacs** [comment]
This is implemented here: https://github.com/unicorn-engine/unicorn/blob/6c1cbef6ac505d355033aef1176b684d02e1eb3a/qemu/target/mips/unicorn.c#L20-L33

Looks like this code tries to support the memory segments of MIPS. But I would say this is not correct implemented because Unicorn only works on virtual addresses and has no "physical memory".

**Comment by wtdcode** [comment]
> This is implemented here:
> 
> https://github.com/unicorn-engine/unicorn/blob/6c1cbef6ac505d355033aef1176b684d02e1eb3a/qemu/target/mips/unicorn.c#L20-L33
> 
> Looks like this code tries to support the memory segments of MIPS. But I would say this is not correct implemented because Unicorn only works on virtual addresses and has no "physical memory".

This is a very old mips dirty hack and I failed to understand it several times. However, for your statement `has no "physical memory"`, it's wrong because `uc_mem_write/read` in fact works with the physical memory.

**Comment by PhilippTakacs** [comment]
I think there is a misunderstanding.

There is no separation between virtual and physical addresses in unicorn (of the emulated target). The MIPS kseg0 and kseg1 required to be mapped the virtual addresses to fixed physical addresses. The current implementation only change the virtual addresses from 0x80000000/0xa0000000 to 0x0 if you want to map/read/write on these addresses. The emulated code then also don't see anything on these virtual addresses.

https://training.mips.com/basic_mips/PDF/Memory_Map.pdf

**Comment by PhilippTakacs** [comment]
This is fixed in e96ac42b2e8f8ded790fd156ef5a5d55ac088acb see #1746 and #1720 for details
